File: backend-javascript/src/auth/api-keys.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { randomBytes, scryptSync, timingSafeEqual } from 'crypto';
import { Repository } from 'typeorm';
import { ApiKey } from '../common/entities/api-key.entity';

@Injectable()
export class ApiKeysService {
  private readonly salt: string;

  constructor(
    @InjectRepository(ApiKey)
    private readonly apiKeysRepository: Repository<ApiKey>
  ) {
    const salt = process.env.API_KEY_SALT || '';
    if (salt.length < 16) {
      throw new Error('API_KEY_SALT must be at least 16 characters');
    }
    this.salt = salt;
  }

  private hash(value: string): string {
    return scryptSync(value, this.salt, 32).toString('hex');
  }

  async create(name: string, tenantId?: string) {
    const plainKey = randomBytes(32).toString('hex');
    const apiKey = this.apiKeysRepository.create({
      name,
      tenantId: tenantId ?? null,
      hashedKey: this.hash(plainKey)
    });
    await this.apiKeysRepository.save(apiKey);

    return {
      id: apiKey.id,
      name: apiKey.name,
      tenantId: apiKey.tenantId,
      apiKey: plainKey
    };
  }

  async list(tenantId?: string) {
    const where = tenantId ? { tenantId } : {};
    return this.apiKeysRepository.find({ where });
  }

  async revoke(id: string) {
    const apiKey = await this.apiKeysRepository.findOne({ where: { id } });
    if (!apiKey) {
      return null;
    }
    apiKey.status = 'revoked';
    return this.apiKeysRepository.save(apiKey);
  }

  async rotate(id: string) {
    const apiKey = await this.apiKeysRepository.findOne({ where: { id } });
    if (!apiKey) {
      return null;
    }
    const plainKey = randomBytes(32).toString('hex');
    apiKey.hashedKey = this.hash(plainKey);
    apiKey.status = 'active';
    await this.apiKeysRepository.save(apiKey);
    return {
      id: apiKey.id,
      name: apiKey.name,
      tenantId: apiKey.tenantId,
      apiKey: plainKey
    };
  }

  async validate(apiKey: string) {
    const computed = this.hash(apiKey);
    const record = await this.apiKeysRepository.findOne({ where: { hashedKey: computed, status: 'active' } });
    if (!record) {
      return null;
    }
    const match = timingSafeEqual(Buffer.from(record.hashedKey, 'hex'), Buffer.from(computed, 'hex'));
    if (!match) {
      return null;
    }
    return record;
  }
}

File: backend-java/src/main/java/com/neria/manager/auth/ApiKeysService.java
package com.neria.manager.auth;

import com.neria.manager.common.entities.ApiKey;
import com.neria.manager.common.repos.ApiKeyRepository;
import com.neria.manager.common.services.ScryptHasher;
import com.neria.manager.config.AppProperties;
import java.security.SecureRandom;
import java.time.LocalDateTime;
import java.util.HexFormat;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import org.springframework.stereotype.Service;

@Service
public class ApiKeysService {
  private final ApiKeyRepository apiKeyRepository;
  private final ScryptHasher hasher;
  private final String salt;

  public ApiKeysService(
      ApiKeyRepository apiKeyRepository, ScryptHasher hasher, AppProperties properties) {
    this.apiKeyRepository = apiKeyRepository;
    this.hasher = hasher;
    this.salt = Optional.ofNullable(properties.getSecurity().getApiKeySalt()).orElse("");
    if (this.salt.length() < 16) {
      throw new IllegalStateException("API_KEY_SALT must be at least 16 characters");
    }
  }

  public ApiKeyCreateResult create(String name, String tenantId) {
    String plainKey = generateKey();
    ApiKey record = new ApiKey();
    record.setId(UUID.randomUUID().toString());
    record.setName(name);
    record.setTenantId(tenantId);
    record.setHashedKey(hasher.hash(plainKey, salt));
    record.setStatus("active");
    record.setCreatedAt(LocalDateTime.now());
    apiKeyRepository.save(record);
    return new ApiKeyCreateResult(record.getId(), record.getName(), record.getTenantId(), plainKey);
  }

  public List<ApiKey> list(String tenantId) {
    if (tenantId != null && !tenantId.isBlank()) {
      return apiKeyRepository.findByTenantId(tenantId);
    }
    return apiKeyRepository.findAll();
  }

  public ApiKey revoke(String id) {
    ApiKey record = apiKeyRepository.findById(id).orElse(null);
    if (record == null) {
      return null;
    }
    record.setStatus("revoked");
    return apiKeyRepository.save(record);
  }

  public ApiKeyCreateResult rotate(String id) {
    ApiKey record = apiKeyRepository.findById(id).orElse(null);
    if (record == null) {
      return null;
    }
    String plainKey = generateKey();
    record.setHashedKey(hasher.hash(plainKey, salt));
    record.setStatus("active");
    apiKeyRepository.save(record);
    return new ApiKeyCreateResult(record.getId(), record.getName(), record.getTenantId(), plainKey);
  }

  public ApiKey validate(String apiKey) {
    String hashed = hasher.hash(apiKey, salt);
    return apiKeyRepository.findByHashedKeyAndStatus(hashed, "active").orElse(null);
  }

  private String generateKey() {
    byte[] bytes = new byte[32];
    new SecureRandom().nextBytes(bytes);
    return HexFormat.of().formatHex(bytes);
  }

  public record ApiKeyCreateResult(String id, String name, String tenantId, String apiKey) {}
}
